[Pipelining] Add schedule unshard/reshard pass

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131762
* #130860
* #130488
* #132369
* #130378
* __->__ #129810

Adds fsdp unshard/reshard ops to a compute-only schedule.

Operates on one pp-rank's schedule at a time, since there is no
cross-pp-rank coordination needed for FSDP.  (Unshard/Reshard is across
DP ranks within a PP group).

Uses a heuristic based on examining the next N stages to run compute
operations on this rank, evicting (resharding) and fetching (unsharding)
ahead of time to give unshard operations a chance to overlap with
compute and PP comms.
- this heuristic has not been validated and may not be optimal

Makes the assumption that it's fine to add the UNSHARD/RESHARD actions
to the schedule regardless of if FSDP will actually be used.
- this way, users do not have to tell us at PP schedule creation time if
  they plan to use FSDP or DDP
- it is trivial to implement UNSHARD/RESHARD as no-ops inside the
  runtime, if FSDP is not detected on the stage module

TODO
- also add FSDP's reduce-scatter? or is it sufficient to leave this
  handled by PipelineStage at 'last backward' time
- validate 'next N stages' heuristic and expose an API if needed
- add an e2e test

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

kwen2501 commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -603,6 +603,135 @@ def _step_microbatches(
+
+    def _per_rank_unshard(rank):
+        rank_actions = compute_actions[rank]
+        active_layers: Set[int] = set()
```
What's the definition of `layer` here? As like a real "layer" in a model or something else? Reads like a concrete thing.

wconstab replied:
yea i was being imprecise. i changed them all to 'stage'/'stages'

H-Huang committed:
Update on "[Draft][Pipelining] Implement compute to comm lowering"


Kind of working pass for adding fsdp unshard/reshard, need to add more
test cases and see if i can improve the logic.

Adding send/recv ops would be a separate 'pass' coming next.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 tianyu-l chauhang d4l3k

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

Skylion007 commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -743,6 +753,135 @@ def _step_microbatches(
+                evict_candidates,
+                num_evict,
+            )
+            for layer in list(evict_candidates)[:num_evict]:
```
nit: you can probably just use itertools islice or something here.

Skylion007 commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -743,6 +753,135 @@ def _step_microbatches(
+            # it thinks the currently active layer is evict-worthy since it doesn't show up in fetch.
+            # it's only saved by num_evict.  But i should figure out a cleaner way.
+            num_evict = max(
+                0, len(set(fetch + list(active_layers))) - max_active_layers
```
itertoools.chain here

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

kwen2501 commented on test/distributed/pipelining/test_schedule.py:
```diff
@@ -115,5 +122,54 @@ def test_pipeline_order(self, ScheduleClass):
+            },
+        ],
+    )
+    def test_unshard_reshard(self, test_info):
```
nit: can you comment on purpose of test for better readability? Thanks!

kwen2501 commented on test/distributed/pipelining/test_schedule.py:
```diff
@@ -10,8 +10,15 @@
+    _add_unshard_reshard,
     _format_pipeline_order,
     _validate_pipeline_order,
+    B,
+    F,
+    RESHARD,
+    UNSHARD,
+    W,
```
nit: can you make B, F, W more descriptive? Can be in a separate PR if it would involve larger changes. Thanks

wconstab replied:
Wdym, name them longer names?  I wanted them to be 1 character so they were easy to read/type in the instantiations below.  I'm happy to add more comments but hopefully you don't insist on writing it out as 'FORWARD' everywhere?

kwen2501 commented on test/distributed/pipelining/test_schedule.py:
```diff
@@ -115,5 +122,54 @@ def test_pipeline_order(self, ScheduleClass):
+            ("3RESHARD", _Action(3, RESHARD)),
+        ],
+    )
+    def test_action_parse(self, action_str_and_ref):
```
nit: can you comment on purpose of test for better readability? Thanks!

kwen2501 commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -53,15 +57,21 @@ def from_str(action):
+RESHARD = _ComputationType.RESHARD
 
-_action_regex = re.compile(r"(\d+)([F,B,W])(\d*)")
+_action_regex = re.compile(r"(\d+)([F,B,W]|UNSHARD|RESHARD)(\d*)")
```
nit: can you comment on what this regex does for better readability? Thanks!

kwen2501 commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -743,6 +753,99 @@ def _step_microbatches(
+
+    def _reshard(stage_index: int):
+        active_stages.remove(stage_index)
+        unshard_actions.append(_Action(stage_index, RESHARD))
```
Why using "unshard_actions" here? Just for saving one variable?

wconstab replied:
not sure what you mean about saving a variable.  unshard_actions is basically the 'new' version of the actions, which includes unshard/reshard.  the original actions list (compute only) is not modified.

kwen2501 commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -743,6 +753,99 @@ def _step_microbatches(
+            evict,
+        )
+
+        for stage in evict:
+            _reshard(stage)
+        for stage in fetch:
+            _unshard(stage)
+        unshard_actions.append(action)
+
+    return unshard_actions
```
A general question: how does this interact with the code of FSDP? Which library would be the actual "firer" of the unshard ops?

wconstab replied:
I am working on a PR now that adds the 'runtime' component that would read this schedule format and execute the operations.  Probably what i'll do for the 'unshard' implementation is, 
 - if the stage module is an FSDP instance, call fsdp's unshard
 - otherwise, no-op.  (this way the unshard lowering can be run in all cases even if someone uses it with DDP..) 

If we need to decide at lowering time whether we're using FSDP vs DDP vs no DP, I think we can do that, but it makes composability a little more annoying so i'd rather not.

kwen2501 reviewed (approved):
Code looks good to me. I only have a general question about interoperability with FSDP and some minor requests for comments.

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

H-Huang commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -36,12 +36,16 @@ class _ComputationType(Enum):
     BACKWARD = 2
     WEIGHT = 3
+    UNSHARD = 4
+    RESHARD = 5
```
High level question naming-wise since the next PR also adds `SEND_F`, `RECV_F`, etc. Should these and unshard/reshard be under a different enum like `_CommunicationType`?

These aren't really related to what the stage is computing and also might be easier to read code-wise to distinguish between the two. And maybe we can refactor the pipeline schedule order's naming to be something like `PipelineScheduleEvents` which encompass the computation and communication

wconstab replied:
> refactor the pipeline schedule order's naming to be something like PipelineScheduleEvents

this sgtm

> Should these and unshard/reshard be under a different enum like _CommunicationType

well in the next PR i initially separated them under _Action.computation_type and _Action.comm_type but I didn't like that and preferred to just have one 'action' field within action.  It should be renamed from _ComputationType to something else though.  Maybe _ActionType.

The reason i like them together instead of separate is becuase in the runtime we basically need a switch statement over the union of them, so having them come from 2 fields is a bit annoying.

I could probably also make _CommType and _CompType and have _Action.action: Union[_CommType, _CompType].  Wdyt?

H-Huang replied:
> I could probably also make _CommType and _CompType and have _Action.action: Union[_CommType, _CompType]. Wdyt?

I like this. I also think this is more applicable later when we want to perform validation on the comms, then you past in the list of the comm types. Or even later if we do automatic bubble estimation, we pass in the list of the comp types. Then the full lowered schedule is a mix of both. I think that is intuitive!

H-Huang commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -743,6 +753,76 @@ def _step_microbatches(
+        return ret
+
+    active_stages: Set[int] = set()
+    unshard_actions: List[Optional[_Action]] = []
```
This naming confused me at first, instead of `unshard_actions` i think this should be renamed like `fsdp_aware_compute_actions` or something

H-Huang commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -743,6 +753,76 @@ def _step_microbatches(
+
+    max_active_stages controls how many prefetches we allow. It should be measured in mb and tuneable but in practice
+    3 stages is probably the thing we want?
+    (to account for having one f and one b active, and something else prefetching?)
```
what did you mean by the "something else prefetching" comment?

wconstab replied:
i'm not sure. i can't even remember why i thought we need 3 instead of 2 now. :/

H-Huang reviewed (approved):
Looks good to me

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge

wconstab committed:
Update

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab committed:
Update

[ghstack-poisoned]
